Add support for vectors on the rhs of linalg.lu_solve

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73878
* #73877
* #74046
* __->__ #74045
* #73806
* #73804
* #73803
* #72935
* #67833

I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

lezcano committed:
Add support for vectors on the rhs of linalg.lu_solve

I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

IvanYashchuk commented:
We don't need `lu_solve` to support vectors for the `B` tensor even if you want in the future to change the internal code of `solve`.

Before adding this "NumPy compatibility" to other functions it shall be discussed (in https://github.com/pytorch/pytorch/issues/52915). For `solve` we need that for NumPy and Array API compatibility. We don't necessarily need this for other "solve" functions.

IvanYashchuk reviewed (changes requested):
See https://github.com/pytorch/pytorch/pull/74045#issuecomment-1112304913.

lezcano committed:
Update on "Add support for vectors on the rhs of linalg.lu_solve"


I really don't like this feature, but I guess that's what we are
supporting. I need it to make `linalg.solve` a composition of two
functions (albeit with custom derivative formulas).

Rant: This PR and the next one in this stack show how annoying and
inconvenient is to support NumPy's behaviour
of allowing the rhs to be a vector. We have to hardcode in every formula
(forward AD / backward AD / the function itself / every test) a few lines to handle
that case. This should be done by the user by calling `.unsqueeze(-1)` before calling
this function if they want to operate on column vectors.

[ghstack-poisoned]

Lezcano commented:
Agree. I'm moving this logic to `linalg.solve` in the next PR, and I'm dropping this PR.
Thank you for the feedback!
